### ISSUE TITLE
Does DetBenchTrain class support nn.DataParallel?

### ISSUE THREAD

**Issue by lognat0704** [rwightman/efficientdet-pytorch#29]
Thanks for the wonderful PyTorch version of EffiDet. I am trying to retrain the EffiDet B5 using multi GPUs, but got much worse loss.(Epoch: 1, summaryl oss: 0.98135) comparing to single GPU (Epoch: 1, summary loss: 0.4458) 

Does the DetBenchTrain supports nn.DataParallel?

`net.class_net = HeadNet(config, num_outputs=config.num_classes, norm_kwargs=dict(eps=.001, momentum=.01))
    return DetBenchTrain(net, config)`

**Comment by GokulEpiphany** [comment]
It does support multi GPUs. If you have Apex installed, DistributedDataParallel from https://nvidia.github.io/apex/parallel.html is used. Else, torch DistributedDataParallel is used. In case of multi GPU setup, I think loss is the summation of n batches being executed parallelly, that might explain 0.98 loss(which is ~0.45*2) assuming you are training on 2 GPUs. This is my understanding of the codebase, I might be wrong. Could you confirm if you are using 2 GPUs? (https://github.com/rwightman/efficientdet-pytorch/blob/master/train.py#L298)

**Comment by rwightman** [comment]
@lognat0704 I have not tested the train bench with DataParallel so a reasonable chance it doesn't work. The reason DataParallel isn't in my train code is that it is slower than DistributedDataParallel. For models like this that take a loong time to train, DDP + SyncBN seems the best approach. In my opinion not worth spending time to fix the slower solution. I do support DP in the validation / predict bench because it's easier to get the completely correct val result (no extra padding for the last batch). 

The loss is a mean, reduced mean across the multiple processes, so it should be the same regardless of mode, if it's 2x there is something wrong. Do keep in mind, you want to scale your LR with the effective batch size, so if you double the effective batch size by doubling the GPUs active, you want to double the LR.

**Comment by glenn-jocher** [comment]
@rwightman we also use torch DDP (have not tried the apex version) for training and DP for val/testing in our yolo repos. We recently had a user claim significant multi gpu performance improvements over torch DDP by using mp spawn for training, though I have not had time to try to verify myself.
